reach

```diff
@@ -115,10 +117,81 @@ run_s () {
   sh -ac "$TMP/out.sh" "$0"
 }
 
+package_manager () {
+  if ( which apt > /dev/null 2>&1 ); then
+    export PACKAGE_MANAGER="apt"
+  fi
+
+  if ( which yum > /dev/null 2>&1 ); then
+    export PACKAGE_MANAGER="yum"
+  fi
+}
+
+install_make () {
+  package_manager
+
+  sudo ${PACKAGE_MANAGER} install make
+}
+
+install_curl () {
+  PACKAGE_MANAGER=package_manager
+
+  sudo $PACKAGE_MANAGER install curl 
+}
+
+install_docker () {
+  WINGET_VERSION="v1.3.431"
+  WINGET_ARTIFACT_NAME="Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle"
+  WINGET_BUNDLE_URL="https://github.com/microsoft/winget-cli/releases/download/${WINGET_VERSION}/${WINGET_ARTIFACT_NAME}"
+  TEMP_DIR="/mnt/c/tmp"
+  WIN_TEMP_DIR="c:\\tmp"
+  DOCKER_BIN="/mnt/c/Program Files/Docker/Docker/resources/bin/docker"
+  PWD=$(pwd)
+
+  cd / #Depending on the directory the user is, windows has a hard time with the path.
+
+  if (which wsl.exe > /dev/null 2>&1); then
+    printf "Installing Winget Windows Package Manager..."
+    if (! which winget.exe > /dev/null 2>&1); then
+      if [ ! -e $TEMP_DIR ]; then
+        mkdir $TEMP_DIR
+      fi
+      sudo wget $WINGET_BUNDLE_URL -O "${TEMP_DIR}/${WINGET_ARTIFACT_NAME}"; powershell.exe "cd ${WIN_TEMP_DIR}; powershell .\\${WINGET_ARTIFACT_NAME}"
+      while ( ! which winget.exe > /dev/null 2>&1 ); do
+        sleep 1;
+      done
+      rm -rf "${TEMP_DIR}/${WINGET_ARTIFACT_NAME}"
+    fi
+    printf "\nInstalling Docker through Winget..."
+    powershell.exe 'Start-Process powershell.exe "winget install docker.dockerdesktop --accept-package-agreements --accept-source-agreements" -Verb runAs'
+    while [ ! -f "${DOCKER_BIN}" ]; do
+      sleep 120;
+    done
+    printf "\nThis installation requires a restart for completion; do you want to restart now? [Y,n]: "
+    read RESPONSE
+    if [ ${RESPONSE} == "Y" ]; then
+      powershell.exe 'shutdown -r -t 0'
+    fi
+  else
+    printf "Make sure you have docker installed following these steps:\n"
+    printf "Ubuntu: https://docs.docker.com/engine/install/ubuntu/\n"
+    printf "Debian: https://docs.docker.com/engine/install/debian/\n"
+    printf "CentOS: https://docs.docker.com/engine/install/centos/\n"
+    printf "Fedora: https://docs.docker.com/engine/install/fedora/\n"
+    printf "RedHat: https://docs.docker.com/engine/install/rhel/\n"
+    printf "SLES: https://docs.docker.com/engine/install/SLES/\n"
+  fi
+  cd $PWD
+}
+
+
 for DEP in make curl docker docker-compose; do
   if ! (which "${DEP}" >/dev/null 2>&1); then
-    echo "Reach relies on an installation of ${DEP}"
-    exit 1
+    printf "Reach relies on an installation of ${DEP}; do you want to install it? [Y,n]: "
+    read RESPONSE
+    if [ ${RESPONSE} == "Y" ]; then
+      eval install_$DEP
```

Comment:
I don't think you should do it like this. I think you should define functions and write the code more directly

Reply:
I was trying to avoid code repetition here.

I will make it more explicit.